drafts/2019-01-21-java-11.adoc

```diff
@@ -0,0 +1,293 @@
+|Most Java EE 7 or 8 features
+
+|java.xml.ws
+v|javax.jws.*
+javax.xml.soap 
+javax.xml.ws.*
+|jaxws-2.2
+|===
+
+=== Introduction of the Java Platform Module System (JPMS)
+
+In Java 9, one of the biggest changes to Java was delivered, under http://openjdk.java.net/jeps/261[JEP 261: The Module System]. This introduces a new Java construct called a "module" which is essentially an archive that is a collection of Java packages, with a `module-info.class` descriptor at the root of the archive. Using the `module-info` descriptor, a module can define what packages it exposes, what services it provides, what services it requires, and what modules it requires.
+
+Here is an example of what a `module-info.java` might look like:
+[source,java]
+----
+module com.foo {
+  // Declare what packages are exported from this module
+  exports com.foo.api;
+  exports com.foo.spi to com.foo.othermodule;
+
+  // Declare a package as "open" so that other modules can perform deep reflection into it
+  // This is useful for CDI injection or JPA entity weaving
+  opens com.foo.internal.jpa.entities;
+
+  // Declare what modules this module requires
+  // By default, every module automatically requires the 'java.base' module
+  requires java.sql;
+  requires java.xml;
+
+  // Declare used and provided services
+  uses java.sql.Driver;
+  provides com.foo.api.FooService
+      with com.foo.internal.FooServiceImpl;
+}
+----
+
+The JDK itself is now organized into Java modules, and it looks something like this:
+
+image::/img/blog/blog_java11_jpms.png[]
+
+The intent is for the entire Java ecosystem to eventually organize itself into Java modules, but for now Java modules are not defined in the JavaEE/JakartaEE/MicroProfile space. *While you can technically include a Java module in your Open Liberty application, the module-info descriptor will have no effect at the moment.*
+
+==== JPMS Access Restrictions
+
+While JPMS does introduce some interesting architectural capabilities and brings a long-needed modularity system into the JDK, it does so at the expense of several popular Java technologies, namely CDI and JPA. The reason for this is that Java modules cannot access members of non-exported packages from other Java modules, and Java code may no longer perform deep reflection (i.e. things that require `setAccessible(true)`) on members of Java modules. If either of these rules are violated, an `IllegalAccessException` is thrown.
+
+Since this is such a major compatibility issue for most of the Java ecosystem, a "kill switch" (formally known as the `--illegal-access=permit` JVM option) has been introduced. It is enabled by default and allows code in unnamed modules to access and perform deep reflection on all named modules. It is important to note that eventually the kill switch will not be enabled by default, and later it will be removed entirely, so it is best to stop writing code that uses deep reflection and start migrating to alternative approaches.
```

Comment:
Even this paragraph makes it sound like the user has control over whether this module access is allowed or not in the future...  But, even a careful developer could still hit problems if they are using a CDI or JPA implementation that still tries to go around these module restrictions.  Or, have our CDI and JPA implementations been all cleared up?  I'm not sure where we are at with running Open Liberty without the "kill switch"...

Reply:
Last I checked, all of our tests still passed even with the kill switch turned off. This is mainly because the illegal access checks do not apply to unnamed modules (which all apps will be until we define JPMS in the EE/MP layer). Even if an application were to include a module-info today, at runtime the app would be running as an unnamed module because Liberty will not load the application as a module.

Eventually, if JPMS is defined in the application space, a developer can always open specific packages, or their entire module. This will allow current JPA/CDI implementations to deep reflect into the "open" code. As far as I'm aware, it will always be required for application developers to open up their code for injection and whatnot.